task.sol

```diff
@@ -0,0 +1,111 @@
+pragma solidity ^0.4.18;
+
+interface IFilesystem {
+
+   function addIPFSFile(string name, uint size, string hash, bytes32 root, uint nonce) external returns (bytes32);
+   function createFileWithContents(string name, uint nonce, bytes32[] arr, uint sz) external returns (bytes32);
+   function getSize(bytes32 id) external view returns (uint);
+   function getRoot(bytes32 id) external view returns (bytes32);
+   function getData(bytes32 id) external view returns (bytes32[]);
+   function forwardData(bytes32 id, address a) external;
+   
+   function makeBundle(uint num) external view returns (bytes32);
+   function addToBundle(bytes32 id, bytes32 file_id) external returns (bytes32);
+   function finalizeBundleIPFS(bytes32 id, string file, bytes32 init) external;
+   function getInitHash(bytes32 bid) external view returns (bytes32);
+   
+   function debug_finalizeBundleIPFS(bytes32 id, string file, bytes32 init) external returns (bytes32, bytes32, bytes32, bytes32, bytes32);
+   
+}
+
+interface ITruebit {
+   function add(bytes32 init, /* CodeType */ uint8 ct, /* Storage */ uint8 cs, string stor) external returns (uint);
+   function addWithParameters(bytes32 init, /* CodeType */ uint8 ct, /* Storage */ uint8 cs, string stor, uint8 stack, uint8 mem, uint8 globals, uint8 table, uint8 call) external returns (uint);
+   function requireFile(uint id, bytes32 hash, /* Storage */ uint8 st) external;
+}
+
+contract Task {
+   uint nonce;
+   ITruebit truebit;
+   IFilesystem filesystem;
+
+   string code;
+   bytes32 init;
+
+   mapping (uint => string) task_to_ipfs;
+   mapping (string => uint) result;
+   
+   event GotFiles(bytes32[] files);
+   event Consuming(bytes32[] arr);
+   
+   constructor(address tb, address fs, string code_address, bytes32 init_hash) public {
+      truebit = ITruebit(tb);
+      filesystem = IFilesystem(fs);
+      code = code_address;     // address for wasm file in IPFS
+      init = init_hash;        // the canonical hash
+   }
+
+   // add new task for a file
+   function submit(string hash, bytes32 root, uint size) public returns (bytes32) {
+      uint num = nonce;
+      nonce++;
+      bytes32 input_file = filesystem.addIPFSFile("input.ts", size, hash, root, num);
+      bytes32 bundle = filesystem.makeBundle(num);
+      filesystem.addToBundle(bundle, input_file);
+      bytes32[] memory empty = new bytes32[](0);
+      filesystem.addToBundle(bundle, filesystem.createFileWithContents("output.data", num+1000000000, empty, 0));
+      filesystem.finalizeBundleIPFS(bundle, code, init);
+      
+      uint task = truebit.addWithParameters(filesystem.getInitHash(bundle), 1, 1, idToString(bundle), 20, 21, 8, 20, 10);
```

Comment:
The identifier of the bundle is `bytes32`, but I use type `string` here because in the current implementation, this argument could also be an ipfs hash. It will probably be changed to just use `bytes32` later